hail/src/test/scala/is/hail/expr/ir/IRSuite.scala

```diff
@@ -1918,6 +1918,72 @@ class IRSuite extends HailSuite {
     assertEvalsTo(makeNDArrayRef(mat2, FastIndexedSeq(0, 0)), 1.0)
   }
 
+  @Test def testNDArrayConcat() {
+    implicit val execStrats = ExecStrategy.compileOnly
+    def nds(ndData: (IndexedSeq[Int], Long, Long)*): IR = {
+      MakeArray(ndData.map { case (values, nRows, nCols) =>
+        if (values == null) NA(TNDArray(TInt32(), Nat(2))) else
+          MakeNDArray(Literal(TArray(TInt32()), values),
+            Literal(TTuple(TInt64(), TInt64()), Row(nRows, nCols)), True())
+      }, TArray(TNDArray(TInt32(), Nat(2))))
+    }
+
+    def assertNDEvals(nd: IR, expected: Array[Array[Int]]): Unit = {
+      val arrayIR = if (expected == null) nd else {
+        val nRows = expected.length
+        val nCols = if (nRows == 0) 0 else expected.head.length
+        Let("nd", nd,
+          MakeArray(
+            Array.tabulate(nRows) { i =>
+              MakeArray(Array.tabulate(nCols) { j =>
+                makeNDArrayRef(Ref("nd", nd.typ), FastIndexedSeq(i, j))
+              }, TArray(TInt32()))
+            }, TArray(TArray(TInt32()))))
+      }
+      assertEvalsTo(arrayIR, if (expected == null) null else expected.map(_.toFastIndexedSeq).toFastIndexedSeq)
+    }
+
+    val nd1 = (FastIndexedSeq(
+      0, 1, 2,
+      3, 4, 5), 2L, 3L)
+
+    val rowwise = (FastIndexedSeq(
+      6, 7, 8,
+      9, 10, 11,
+      12, 13, 14), 3L, 3L)
+
+    val colwise = (FastIndexedSeq(
+      15, 16,
+      17, 18), 2L, 2L)
+
+    val emptyRowwise = (FastIndexedSeq(), 0L, 3L)
+    val emptyColwise = (FastIndexedSeq(), 2L, 0L)
+    val na = (null, 0L, 0L)
+
+    val rowwiseExpected = Array(
+      Array(0, 1, 2),
+      Array(3, 4, 5),
+      Array(6, 7, 8),
+      Array(9, 10, 11),
+      Array(12, 13, 14))
+    val colwiseExpected = Array(
+      Array(0, 1, 2, 15, 16),
+      Array(3, 4, 5, 17, 18))
+
+    assertNDEvals(NDArrayConcat(nds(nd1, rowwise), 0), rowwiseExpected)
+    assertNDEvals(NDArrayConcat(nds(nd1, rowwise, emptyRowwise), 0), rowwiseExpected)
+    assertNDEvals(NDArrayConcat(nds(nd1, emptyRowwise, rowwise), 0), rowwiseExpected)
+
+    assertNDEvals(NDArrayConcat(nds(nd1, colwise), 1), colwiseExpected)
+    assertNDEvals(NDArrayConcat(nds(nd1, colwise, emptyColwise), 1), colwiseExpected)
+    assertNDEvals(NDArrayConcat(nds(nd1, emptyColwise, colwise), 1), colwiseExpected)
+
+    // FIXME: These are changing type during PruneDeadFields for some reason...
```

Comment:
This is almost definitely some requiredness bullshit. I ran into something similar on my linear regression rows branch,but I've just turned off `PruneDeadFields` on that branch since I didn't want to go down a different rabbit hole.

Reply:
yeah, it is---the type change is from `TArray(TNDArray(TInt32(), 2))` to `TArray(TNDArray(+TInt32(), 2))`. 🙃